iocage/lib/Network.py

```diff
@@ -21,75 +21,106 @@
-        self.bridges = bridges
+        self.bridge = bridge
         self.jail = jail
         self.nic = nic
         self.mtu = mtu
         self.ipv4_addresses = ipv4_addresses or []
         self.ipv6_addresses = ipv6_addresses or []
 
-    def setup(self):
-        if self.vnet:
-            if not self.bridges or len(self.bridges) == 0:
-                raise iocage.lib.errors.VnetBridgeMissing(
-                    logger=self.logger
-                )
+    def setup(self) -> None:
+        if (self.vnet is True) and (self.bridge is None):
+            raise iocage.lib.errors.VnetBridgeMissing(logger=self.logger)
+        self.__create_vnet_iface()
+
+    def teardown(self) -> None:
+        if self.vnet is True:
+            self.__down_host_interface()
+            if self.bridge.secure is True:
+                self.__down_secure_mode_devices()
+
+    def __down_host_interface(self) -> None:
+        iocage.lib.NetworkInterface.NetworkInterface(
+            name=self.nic_local_name,
+            extra_settings=["destroy"],
+            logger=self.logger
+        )
+
+    def __down_secure_mode_devices(self) -> None:
+
+        self.logger.verbose("Downing secure mode devices")
 
-            jail_if, host_if = self.__create_vnet_iface()
+        for nic in [
+            f"{self.nic_local_name}:a",
+            f"{self.nic_local_name}:net"
+        ]:
 
-    def teardown(self):
-        if self.vnet:
-            # down host_if
             iocage.lib.NetworkInterface.NetworkInterface(
-                name=self.nic_local_name,
+                name=nic,
                 extra_settings=["destroy"],
                 logger=self.logger
             )
 
     @property
-    def nic_local_name(self):
+    def nic_local_name(self) -> str:
         self.jail.require_jail_running(silent=True)
         return f"{self.nic}:{self.jail.jid}"
 
     @property
-    def nic_local_description(self):
-        return f"associated with jail: {self.jail.humanreadable_name}"
+    def nic_group_name(self) -> str:
+        self.jail.require_jail_running(silent=True)
+        return f"ioc-{self.jail.jid}"
 
-    def __create_vnet_iface(self):
+    @property
+    def nic_local_description(self) -> str:
+        return f"associated with jail: {self.jail.humanreadable_name}"
 
-        # create new epair interface
-        epair_a_cmd = ["/sbin/ifconfig", "epair", "create"]
-        epair_a = subprocess.Popen(  # nosec: TODO: use exec helper?
-            epair_a_cmd, stdout=subprocess.PIPE, shell=False).communicate()[0]
-        epair_a = epair_a.decode("utf-8").strip()
-        epair_b = f"{epair_a[:-1]}b"
+    def __create_new_epair_interface(self) -> typing.Tuple[str, str]:
+        epair_a = iocage.lib.NetworkInterface.NetworkInterface(
+            name="epair",
+            create=True,
+            group=self.nic_group_name,
+            logger=self.logger
+        )
+        epair_a_name = epair_a.name
+        epair_b_name = f"{epair_a_name[:-1]}b"
```

Comment:
what does this mean?

Reply:
Remove the last character and add a `b` instead. See the [epair man page](https://www.freebsd.org/cgi/man.cgi?query=epair&sektion=4&manpath=FreeBSD+11.1-RELEASE).